src/hooks/useLastAccessedReportID.ts

```diff
@@ -0,0 +1,86 @@
+ */
+
+let reports: OnyxCollection<Report> = {};
+let reportMetadata: OnyxCollection<ReportMetadata> = {};
+let policies: OnyxCollection<Policy> = {};
+let accountID: number | undefined;
+let isFirstTimeNewExpensifyUser = false;
+
+function subscribeToOnyxData() {
+    // eslint-disable-next-line rulesdir/prefer-onyx-connect-in-libs
+    const reportsConnection = Onyx.connect({
+        key: ONYXKEYS.COLLECTION.REPORT,
+        waitForCollectionCallback: true,
+        callback: (value) => (reports = value),
+    });
+    // eslint-disable-next-line rulesdir/prefer-onyx-connect-in-libs
+    const reportMetadataConnection = Onyx.connect({
+        key: ONYXKEYS.COLLECTION.REPORT_METADATA,
+        waitForCollectionCallback: true,
+        callback: (value) => (reportMetadata = value),
+    });
+    // eslint-disable-next-line rulesdir/prefer-onyx-connect-in-libs
+    const policiesConnection = Onyx.connect({
+        key: ONYXKEYS.COLLECTION.POLICY,
+        waitForCollectionCallback: true,
+        callback: (value) => (policies = value),
+    });
+    // eslint-disable-next-line rulesdir/prefer-onyx-connect-in-libs
+    const isFirstTimeNewExpensifyUserConnection = Onyx.connect({
+        key: ONYXKEYS.NVP_IS_FIRST_TIME_NEW_EXPENSIFY_USER,
+        callback: (value) => (isFirstTimeNewExpensifyUser = !!value),
+    });
+    // eslint-disable-next-line rulesdir/prefer-onyx-connect-in-libs
+    const accountIDConnection = Onyx.connect({
+        key: ONYXKEYS.SESSION,
+        callback: (value) => (accountID = value?.accountID),
+    });
+    return () => {
+        Onyx.disconnect(reportsConnection);
+        Onyx.disconnect(reportMetadataConnection);
+        Onyx.disconnect(policiesConnection);
+        Onyx.disconnect(isFirstTimeNewExpensifyUserConnection);
+        Onyx.disconnect(accountIDConnection);
+    };
+}
+
+/**
+ * Get the last accessed reportID.
+ */
+export default function useLastAccessedReportID(shouldOpenOnAdminRoom: boolean) {
+    const {canUseDefaultRooms} = usePermissions();
+    const {activeWorkspaceID} = useActiveWorkspace();
+
+    const getSnapshot = useCallback(() => {
+        const policyMemberAccountIDs = getPolicyEmployeeListByIdWithoutCurrentUser(policies, activeWorkspaceID, accountID);
+        return ReportUtils.findLastAccessedReport(
+            reports,
+            !canUseDefaultRooms,
+            policies,
+            isFirstTimeNewExpensifyUser,
+            shouldOpenOnAdminRoom,
+            reportMetadata,
+            activeWorkspaceID,
+            policyMemberAccountIDs,
+        )?.reportID;
+    }, [activeWorkspaceID, canUseDefaultRooms, shouldOpenOnAdminRoom]);
+
+    // We need access to all the data from these useOnyx calls, but we don't want to re-render the consuming component
+    // unless the derived value (lastAccessedReportID) changes. To address these, we'll wrap everything with
```

Comment:
I guess it could be omitted but I think its better to leave it in the comment too
```suggestion
    // unless the derived value (lastAccessedReportID) changes. To address these, we'll wrap everything with useSyncExternalStore
```